syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,988 @@
+    must specify the Spanner database URI string to connect to, as follows:
+
+        spanner://INSTANCE_ID:DATABASE_ID
+
+    """
+
+    def __init__(self, sqluri, standard_collections=False, **dbkwds):
+        self.sqluri = sqluri
+        instance_id, database_id = sqluri[len("spanner://"):].split(":")
+        self.client = spanner.Client()
+        self._instance = self.client.instance(instance_id)
+        self._pool = pool = spanner.BurstyPool()
+        self._database = self._instance.database(database_id, pool=pool)
+
+        # There doesn't seem to be a reliable cross-database way to set the
+        # initial value of an autoincrement column.
+        self.standard_collections = standard_collections
+        if self.standard_collections and dbkwds.get("create_tables", False):
+            pass
+
+        # A local in-memory cache for the name => collectionid mapping.
+        self._collections_by_name = {}
+        self._collections_by_id = {}
+        if self.standard_collections:
+            for id, name in STANDARD_COLLECTIONS.iteritems():
+                self._collections_by_name[name] = id
+                self._collections_by_id[id] = name
+
+        # A thread-local to track active sessions.
+        self._tldata = threading.local()
+
+    def _lock_write_precondition(self, session, user, collection):
+        """'Lock' the precondition on the collection
+
+        Get the timestamp of the collection in use, utilizing it as
+        precondition for
+        transaction invalidation.
+
+        """
+        result = session.transaction.execute_sql(
+            COLLECTION_CURRENT_TIMESTAMP,
+            params={"userid": user_key(user),
+                    "collectionid": collection},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64}
+        ).one_or_none()
+        if result:
+            current_ts, last_modified = result
+            session.collection_ts = dt2ts(last_modified)
+            session.timestamp = dt2ts(current_ts)
+            # Forbid the write if it would not properly incr the timestamp.
+            if session.collection_ts >= dt2ts(current_ts):
+                msg = "last_modified: %s (%s) current_ts: %s (%s)" % (
+                    last_modified,
+                    session.collection_ts,
+                    current_ts,
+                    dt2ts(current_ts)
+                )
+                raise ConflictError("lock_for_write %s" % msg)
+        else:
+            result = session.transaction.execute_sql(
+                "SELECT CURRENT_TIMESTAMP()"
+            ).one()[0]
+            session.timestamp = dt2ts(result)
+            session.collection_ts = 0
+
+    @contextlib.contextmanager
+    def lock_for_read(self, ruser, collection):
+        with SpannerStorageSession(self) as spanner_session:
+            with self._database.snapshot(multi_use=True) as txn:
```

Comment:
Snapshots feel nicer for this than a "read lock" 👍

Reply:
Yea, in Spanner its effectively a read-lock, in that all the reads have a single period of consistency.